File: BE/src/model/calendar.model.ts
import mongoose, { Schema, Document } from "mongoose";

export enum CalendarStatus {
  NOT_YET = "not_yet",
  IN_PROGRESS = "in_progress",
  COMPLETED = "completed",
}

export enum AttendanceStatus {
  NOT_YET = "not_yet",
  ABSENT = "absent",
  PRESENT = "present",
}

export interface ICourseCalendar extends Document {
  courseId: mongoose.Types.ObjectId;
  sessionId: mongoose.Types.ObjectId;
  teacherId: mongoose.Types.ObjectId;
  date: Date;
  note?: string;
  status: CalendarStatus;
  createdAt: Date;
  updatedAt: Date;
  attendances: {
    _id?: mongoose.Types.ObjectId;
    userId: mongoose.Types.ObjectId;
    status: AttendanceStatus;
  }[];
}

const CourseCalendarSchema = new Schema<ICourseCalendar>(
  {
    courseId: { type: Schema.Types.ObjectId, ref: "Course", required: true },
    sessionId: { type: Schema.Types.ObjectId, ref: "Session", required: true },
    teacherId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    date: { type: Date, required: true },
    note: { type: String },
    status: {
      type: String,
      enum: Object.values(CalendarStatus),
      default: CalendarStatus.NOT_YET,
    },
    attendances: [
      {
        userId: { type: Schema.Types.ObjectId, ref: "User", required: true },
        status: {
          type: String,
          enum: Object.values(AttendanceStatus),
          default: AttendanceStatus.NOT_YET,
        },
      },
    ],
  },
  { timestamps: true }
);

export const CourseCalendar = mongoose.model<ICourseCalendar>(
  "CourseCalendar",
  CourseCalendarSchema
);

File: mirai_mobile/lib/models/calendar_model.dart
class CalendarModel {
  final String id;
  final dynamic courseId; 
  final String date;
  final dynamic sessionId;
  final dynamic teacherId; 
  final String status;
  final String? note;
  final String? createdAt;
  final String? updatedAt;

  String get courseName => courseId is Map ? (courseId['name'] ?? courseId['courseName'] ?? '') : '';
  String get sessionName => sessionId is Map ? (sessionId['sessionName'] ?? '') : '';
  String get startTime => sessionId is Map ? (sessionId['startTime'] ?? '') : '';
  String get endTime => sessionId is Map ? (sessionId['endTime'] ?? '') : '';
  String get teacherName => teacherId is Map ? (teacherId['name'] ?? teacherId['fullName'] ?? '') : '';

  CalendarModel({
    required this.id,
    this.courseId,
    required this.date,
    this.sessionId,
    this.teacherId,
    required this.status,
    this.note,
    this.createdAt,
    this.updatedAt,
  });

  factory CalendarModel.fromJson(Map<String, dynamic> json) {
    return CalendarModel(
      id: json['_id'] ?? json['calendarId'] ?? '',
      courseId: json['courseId'],
      date: json['date'] ?? '',
      sessionId: json['sessionId'],
      teacherId: json['teacherId'],
      status: json['status'] ?? 'not_yet',
      note: json['note'],
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
    );
  }
}
